src/QsCompiler/DataStructures/SyntaxTree.fs

```diff
@@ -671,6 +673,37 @@ and QsStatement =
     }
 
 
+/// The source files for a syntax tree node.
+type Source =
+    {
+        /// The path to the original source code file.
+        CodeFile: string
+
+        /// The path to the assembly file if the node was loaded from a reference.
+        AssemblyFile: string QsNullable
+    }
+
+    /// The assembly file path for this source if one exists, otherwise the code file path.
+    member source.AssemblyOrCode = source.AssemblyFile.ValueOr source.CodeFile
+
+    /// <summary>
+    /// Returns a copy of this source with the given <paramref name="codeFile"/> or <paramref name="assemblyFile"/> if
+    /// provided.
+    /// </summary>
+    member source.With([<Optional; DefaultParameterValue null>] ?codeFile,
+                       [<Optional; DefaultParameterValue null>] ?assemblyFile) =
+        { source with
+            CodeFile = codeFile |> Option.defaultValue source.CodeFile
+            AssemblyFile = assemblyFile |> QsNullable<_>.FromOption |> QsNullable.orElse source.AssemblyFile
+        }
+
+/// Operations for source files.
+module Source =
+    /// The assembly file path for this source if one exists, otherwise the code file path.
+    [<CompiledName "AssemblyOrCode">]
```

Comment:
While reading the code that uses `AssemblyOrCode`, I originally interpreted it as getting the path to the assembly or getting the code itself. I think it would help readability to be more explicit in the name, e.g. `AssemblyFileOrCodeFile`.

Reply:
How about `AssemblyOrCodeFile`?

Reply:
`AssemblyOrCodeFile` sounds good :)